pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -928,6 +943,27 @@ public void redeliverUnacknowledgedMessages(List<MessageIdImpl> messageIds) {
         }
     }
 
+    private int removeExpiredMessagesFromQueue(Set<MessageIdImpl> messageIds) {
+        int messagesFromQueue = 0;
+        for (Message message : incomingMessages) {
+            MessageIdImpl messageId = (MessageIdImpl) message.getMessageId();
+            if (messageId instanceof BatchMessageIdImpl) {
+                // messageIds contain MessageIdImpl, not BatchMessageIdImpl
+                messageId = new MessageIdImpl(messageId.getLedgerId(), messageId.getEntryId(),
+                        getPartitionIndex());
+            }
+            if (messageIds.contains(messageId)) {
```

Comment:
Another thought: all messages in the queue should timeout in order.. so couldn't we just keep popping elements from the head of the queue until we find one that hasn't expired? 
Eg: (in pseudo-code) 

```java
while (true) {
    T item = queue.peek();
     if (item == null || !item.isExpired()) {
         break;
     }
    
     queue.pop();
}

```